<b>Format Example (for human assembly NCBI version 36): SPACE BASED COORDINATE SYSTEM (comma separated)</b>
<HR>
10,102760304,102760304,1,GCGGCT,#User comment 1
<br>
10,50205013,50205013,1,ACACACACACAC
<br>
5,179134934,179134935,1,/,#User comment 2
<br>
1,153108866,153108866,1,CTGCTGCTGCTG
<br>
11,6368547,6368547,1,GCTGGCGCTGGC
<br>
11,65081932,65081932,1,AGCAGC
<br>
12,110521161,110521164,1,/
<br>
12,116990733,116990736,1,/
<br>
12,123453048,123453048,1,CTG
<br>
12,131113090,131113090,1,GCA
<br>
12,1932613,1932613,1,CTG
<br>
<br>
<BR>
<b>Format Description</b>
<HR>
[comma separated: chromosome,coordinate,oientation,alleles,user comment(optional) ]<BR>
Please do not use spaces except in the user comments field<br>
<br><i>Coordinate System:</i><BR>
SIFT accepts only space-based coordinates for insertion / deletion variants.<BR>  
The space-based coordinate system counts the spaces before and after bases rather than the bases themselves.<BR>
Zero always refers to the space before the first base.<BR><BR> 
The sequence 'ACGT' has coordinates (0,4) and its subsequence 'CG' has coordinates (1,3) as shown in Example 1 below.<BR> 
The difference between the start and end coordinates gives the sequence length. Misinterpretation of these<BR> 
coordinates can easily lead to 'off-by-one' errors.  Space-based coordinates become necessary when describing<BR>
insertions/deletions and genomic rearrangements.<BR><BR>
Example 1:<BR>
<p><font size="1">0</font> A <font size="1">1</font> C <font size="1">2</font> G <font size="1">3</font>  T <font size="1">4</font></p><hr>
<i>Orientation:</i><BR>
Use 1 for positive strand and -1 for negative strand. If  orientation is not known, use 1 as default.<BR><HR>
<BR>
<i>Alleles:</i><BR>
For Insertion, the begin and end coordinates should be same and the allele should be a string of inserted nucleotides in one of the following formats.<br>
1. ----/ATGC<br>
2. -/ATGC<br>
3. ATGC<br>
<br> 
For Deletion, the difference between begin and end coordinates should be equal to the length of the deleted string. the allele can either be left blank or be specified in one of the followig formats<br>
1. ATGC/----<br>
2. ATGC/-<br>
3. /<br>

<HR><br>
<b>Output Description</b>
<HR>
<br>
<i><a name="Amino_acid_position_change"><b><u>Amino Acid Position Change</u></b></a></i><br><br>
This column contains the change coordinates within the original protein sequence and the modified protein sequence. For example, the insertion of GCGGCT at location 102760304 of chromosome 10 of Homo Sapiens (represented by input row: 0,102760304,102760304,1,GCGGCT) inserts two additional amino acids Arginine 'R' and Serine 'S' at position 145 to 147 (space based coordinates) in the modified protein sequence. The mismatch between the original and modified sequences is displayed in red.
<PRE> 
>ENST00000238965; MISMATCH = 145-145
GPQEQGSPASCFETSPAGHATQASPYHPRACRGGFYLLPVNGFPEEEDNGELRERLGALK
VSPSASAPRHPHKGIPPLQDVPVDAFTPLRIACTPPPQLPPVAPRPLRPNWLLTEPLSRE
HPPQSQIRGRAQSRSRSRSRSRSRSSRGQGKSPGRRSPSPVPTPAPSMTNGRYHKPRKAR
PPLPRPLDGEAAKVGAKQGPSESGTEGTAKEAAMKNPSGELKTVTLSKMKQSLGISISGG
IESKVQPMVKIEKIFPGGAAFLSGALQAGFELVAVDGENLEQVTHQRAVDTIRRAYRNKA
REPMELVVRVPGPSPRPSPSDSSALTDGGLPADHLPAHQPLDAAPVPAHWLPEPPTNPQT
PPTDARLLQPTPSPAPSPALQTPDSKPAPSPRIP
 
>ENST00000238965; MISMATCH = 145-147
GPQEQGSPASCFETSPAGHATQASPYHPRACRGGFYLLPVNGFPEEEDNGELRERLGALK
VSPSASAPRHPHKGIPPLQDVPVDAFTPLRIACTPPPQLPPVAPRPLRPNWLLTEPLSRE
HPPQSQIRGRAQSRSRSRSRSRSRS<font color=red>rs</font>SRGQGKSPGRRSPSPVPTPAPSMTNGRYHKPRK
ARPPLPRPLDGEAAKVGAKQGPSESGTEGTAKEAAMKNPSGELKTVTLSKMKQSLGISIS
GGIESKVQPMVKIEKIFPGGAAFLSGALQAGFELVAVDGENLEQVTHQRAVDTIRRAYRN
KAREPMELVVRVPGPSPRPSPSDSSALTDGGLPADHLPAHQPLDAAPVPAHWLPEPPTNP
QTPPTDARLLQPTPSPAPSPALQTPDSKPAPSPRIP
</PRE> 
<br>
<HR>
<i><a name="Indel_location"><u><b>Indel location</b></u></a></i><br><br>
This percentage indicates the approximate location of the indel in the protein. 
 For example, a value less than 50% means that the indel is located in the first half of the protein and is close to the amino terminus, whereas a number greater than 50% means that the indel is closer to the carboxy terminus.
<br><br>

<img src="/www/pics/indel_location.png"/><br>
<BR>
On the x-axis is the relative protein location of the coding indel, which is the first amino acid position of the indel divided by the protein length. A relative protein location near zero indicates that the indel is located near the N-terminus of the protein and a relative protein location near one indicates that the indel is located near the C-terminus of the protein. 
<BR>
Plotted are coding indels found in a human with no major diseases. The coding indels occurr frequently at the N- and C-termini of proteins. These may be functionally neutral because for indels at the end of the protein, most of the protein has already been translated.  Indels at the beginning of the protein may be rescued by a downstream translation start site.  Hence, using the indel location may help you distinguish between neutral and deleterious indels.<br>
[<a href="http://www.plosgenetics.org/article/info:doi/10.1371/journal.pgen.1000160" target=_blank> Reference</a>]

<br><br>
<HR>
<b><i><u>Transcript Visualization</i></u></b><br>
<BR>
Placing the cursor over the contents of this column displays a visualization of the transcript in the following format: 
<br>
<br>
<---{}--{}[]--[*.]--[]--[]--[]--[]--[]--[]--[]{}---|
<br><br>
The above example visualization mimics the structure of the transcript containing the indel. <br>
<PRE>
<--- indicates the 3' end
---| indicates the 5' end
{}   indicate UTR 
[]   indicates a coding exon
--   indicats an intron
.    indicates the start of insertion or deletion
*    indicates the end of deletion<br>
If the 3'end of the transcript appears to the left of the 5' end, as in this case, then the transcript is transcribed from the negative strand.
This transcript has two 3'UTRs, one 5'UTR, nine exons and nine introns.
The indel both starts and ends in the 8th coding exon. 
</PRE>
<br>
<HR>
<i><a name="Nucleotide_change"><b><u>Nucleotide change</a></i></b></u><br>
<br>
The input allele (insertion or deletion) and +/- 5 base pairs are shown. For example, <br>
the user input for insertion variant "10,102760304,102760304,1,GCGGCT" will populate this column with the following information<br>
cggct-<font color="red">GCGGCT</font>-acggc<br><br>
whereas a user input for deletion variant "12,110521161,110521164,1,/" will populate this column with the following information<br>
TGCTG-<font color="red">ctg</font>-TTGCT<br><br>
The indel is displayed in red. For insertions, the inserted bases are displayed in uppercase and the flanking bases are displayed in lowercase.  For deletions, the deleted bases are displayed in lowercase whereas the flanking bases are displayed in uppercase.
<br><br><br>
<HR>
<i><a name="Amino_acid_change"><b><u>Amino acid change</a></i></b></u><br>
<br>
This column displays the amino acid change caused by the indel. For example<br>
QQTT->QQ<font color="red">q</font>TT indicates the addition of amino acid Glutamine ('Q') in the modified protein sequence,<br>
whereas EE<font color="red">e</font>DA->EEDA indicates the deletion of amino acid Glutamic acid, 'E' in the modified protein sequence.<br>
<br><br>
<HR>
<i><a name="Protein_sequence_change"><b><u>Protein sequence change</a></b></u></i><br>
<br>
This column links to page that contains the original and modified protein sequence files with regions of mismatch (caused due to indel) colored in red. For example, an insertion represented by the user input "1,153108866,153108866,1,CTGCTGCTGCTG" causes an expansion in polyglutamine tract as shown in the following fasta format sequences. The Fasta headers contain the Ensembl transcript ID along with the coordinates of change.
<PRE> 
>ENST00000271915; MISMATCH = 80-80
MDTSGHFHDSGVGDLDEDPKCPCPSSGDEQQQQQQQQQQQQPPPPAPPAAPQQPLGPSLQ
PQPPQLQQQQQQQQQQQQQQPPHPLSQLAQLQSQPVHPGLLHSSPTAFRAPPSSNSTAIL
HPSSRQGSQLNLNDHLLGHSPSSTATSGPGGGSRHRQASPLVHRRDSNPFTEIAMSSCKY
SGGVMKPLSRLSASRRNLIEAETEGQPLQLFSPSNPPEIVISSREDNHAHQTLLHHPNAT
HNHQHAGTTASSTTFPKANKRKNQNIGYKLGHRRALFEKRKRLSDYALIFGMFGIVVMVI
ETELSWGLYSKDSMFSLALKCLISLSTIILLGLIIAYHTREVQLFVIDNGADDWRIAMTY
ERILYISLEMLVCAIHPIPGEYKFFWTARLAFSYTPSRAEADVDIILSIPMFLRLYLIAR
VMLLHSKLFTDASSRSIGALNKINFNTRFVMKTLMTICPGTVLLVFSISLWIIAAWTVRV
CERYHDQQDVTSNFLGAMWLISITFLSIGYGDMVPHTYCGKGVCLLTGIMGAGCTALVVA
VVARKLELTKAEKHVHNFMMDTQLTKRIKNAAANVLRETWLIYKHTKLLKKIDHAKVRKH
QRKFLQAIHQLRSVKMEQRKLSDQANTLVDLSKMQNVMYDLITELNDRSEDLEKQIGSLE
SKLEHLTASFNSLPLLIADTLRQQQQQLLSAIIEARGVSVAVGTTHTPISDSPIGVSSTS
FPTPYTSSSSC
 
>ENST00000271915; MISMATCH = 80-84
MDTSGHFHDSGVGDLDEDPKCPCPSSGDEQQQQQQQQQQQQPPPPAPPAAPQQPLGPSLQ
PQPPQLQQQQQQQQQQQQQQ<font color=red>qqqq</font>PPHPLSQLAQLQSQPVHPGLLHSSPTAFRAPPSSNS
TAILHPSSRQGSQLNLNDHLLGHSPSSTATSGPGGGSRHRQASPLVHRRDSNPFTEIAMS
SCKYSGGVMKPLSRLSASRRNLIEAETEGQPLQLFSPSNPPEIVISSREDNHAHQTLLHH
PNATHNHQHAGTTASSTTFPKANKRKNQNIGYKLGHRRALFEKRKRLSDYALIFGMFGIV
VMVIETELSWGLYSKDSMFSLALKCLISLSTIILLGLIIAYHTREVQLFVIDNGADDWRI
AMTYERILYISLEMLVCAIHPIPGEYKFFWTARLAFSYTPSRAEADVDIILSIPMFLRLY
LIARVMLLHSKLFTDASSRSIGALNKINFNTRFVMKTLMTICPGTVLLVFSISLWIIAAW
TVRVCERYHDQQDVTSNFLGAMWLISITFLSIGYGDMVPHTYCGKGVCLLTGIMGAGCTA
LVVAVVARKLELTKAEKHVHNFMMDTQLTKRIKNAAANVLRETWLIYKHTKLLKKIDHAK
VRKHQRKFLQAIHQLRSVKMEQRKLSDQANTLVDLSKMQNVMYDLITELNDRSEDLEKQI
GSLESKLEHLTASFNSLPLLIADTLRQQQQQLLSAIIEARGVSVAVGTTHTPISDSPIGV
SSTSFPTPYTSSSSC
</PRE> 

<br>
<HR>
<i><a name="NMD"><b><u>Causes Nonsense Mediated Decay</a></i></b></u><br>
<br>Nonsense mediated decay (NMD) is a cellular mechanism of mRNA surveillance to detect nonsense mutations and prevent the expression of truncated or erroneous proteins <a href="http://en.wikipedia.org/wiki/Nonsense_mediated_decay">[More]</a>.   <br>
This column indicates whether the input indel is likely to cause NMD.  If NMD occurs, then the indel is equivalent to a gene deletion because the mRNA is never translated. <br>
<br>
There is <b>no</b> NMD when: <BR>
1) the resulting premature termination codon is in the last exon<BR>
-or-<BR>
2) the resulting premature termintion codon is in the last 50 nucleotides in the second to last exon

<br>[<A HREF="http://www.ncbi.nlm.nih.gov/sites/entrez?db=pubmed&cmd=search&term=9644970">Reference</A HREF>]
<br>
<br>
<img src="/www/pics/NMD.png"/><br>
<br><br>
<HR>
<i><a name="repeat"><b><u>Repeat detected</a></u></b></i><br>
<br>This column gets populated if the input insertion/deletion is found to expand or contract a coding repeat region. For example, an input row '1,153108866,153108866,1,CTGCTGCTGCTG' causes an insertion resulting in the expansion of a poly-glutamine repeat. A poly-glutamine repeat of length 14 that expands to length 18 is illustrated in this column by 'PQL(q)14P-->PQL(q)18P'. The repeat amino acid(s) are shown in parenthesis followed by the repeat number and bounded by flanking amino acids.
<br><br>
<HR>
<i><a name="miscall"><b><u>Warning: NCBI reference miscall</a></b></u></i><br>
<br>
If you receive a reference miscall warning in the coordinates column (first column) of the output table, this means that your input coordinates overlap or contain a location that is not a true indel, but likely to be an error in NCBI human genome reference sequence. 

[<a href="http://www.nature.com/nature/journal/vaop/ncurrent/full/nature08211.html" target=_blank>Reference</a><font size ="2">]


<HR>

