454
<BR>
Here is a example line of the SNP calling output, and what each section means:
<PRE>0 3076 G A 0.0194% 15376/ 66 acc (335/ 8 rej): A: 3/ 0 ( 0/ 0) C: 5/ 0 ( 1/ 0) G: 15367/ 66 (334/ 8) T: 1/ 0 ( 0/ 0) iA: 0/ 0 iC: 0/ 0 iG: 0/ 0 iT: 0/ 0 Del: 0/0 F</PRE>
<BR>
0= reference number
<BR>
3076= reference base (dash means deletion)
<BR>
G= nucleotide in reference (dash means deletion)
<BR>
A= nucleotide in sample (454 data)
<BR>
0.0194%= percent of reads disagreeing with the reference
<BR>
15376/ 66 acc= total forward/total reverse reads accepted
<BR>
(335/ 8 rej):= total forward/total reverse reads rejected
<BR>
A: 3/ 0 ( 0/ 0)= total forward reads calling A/ total reverese reads call A accepted (rejected reads calling A: fw/rc)
<BR>
C: 5/ 0 ( 1/ 0)= same as above for C calls
<BR>
G: 15367/ 66 (334/ 8)= same as above for G calls
<BR>
T: 1/ 0 ( 0/ 0)= same as above for T calls
<BR>
iA: 0/ 0= forward/reverse reads indicating an A inserted after this position
<BR>
iC: 0/ 0= same as above for C insertions
<BR>
iG: 0/ 0= same as above for G insertions
<BR>
iT: 0/ 0= same as above for T insertions
<BR>
Del: 0/0= forward/reverse reads indicating a deletion of this base
<BR>
F= true or false. This is only informative if a truth file is given to the program
<BR>
<BR>
*********************
<BR>
Solid
<BR>
<PRE># cov	ref	consen	score	confi	F3	R3	score	conf	F3	R3	coord</PRE>
<PRE>94	C	M	0.6002	0.9108	26/15	30/14	0.2598	0.8830	10/6	15/4	3429197</PRE>
<PRE>11	T	Y	0.4650	0.9307	4/3	1/1	0.2736	0.9126	0/0	3/2	585338</PRE>
<BR>

*********************
