<body bgcolor=white>
<TITLE>SIFT Help</TITLE>
<H1>
<p><center><img src="pics/sifter.gif"  height=95 width=95>
<p><BR>
S<font color=red>I</font>F<font color=blue>T</font> Help</center>
</H1>
<UL>
<LI><A HREF="#SIFT">What does SIFT do?</A>
<LI><A HREF="#SIFT_FOR_ME">What can SIFT do for me?</A>
<LI><A HREF="#SIFT_PROCEDURE">How does SIFT work?</A>
<LI><A HREF="#SIFT_INPUT">Input for SIFT</A>
	<UL>
	<LI><A HREF="#SIFT_INPUT_BLINK">Submitting a NCBI GI #</A>
	<LI><A HREF="#SIFT_INPUT_SEQUENCE">Submitting a sequence</A>
	<LI><A HREF="#SIFT_INPUT_SEQUENCES">Submitting a group of related sequences</A>
	<LI><A HREF="#SIFT_INPUT_ALIGNMENT">Submitting a multiple alignment</A>
	</UL>
<LI><A HREF="#CHANGING_PARAMETERS">Changing SIFT parameters</A>
<LI><A HREF="#SIFT_OUTPUT">SIFT Output</A>
	<UL>
	<LI><A HREF="#SIFT_OUTPUT_SUBST">Predictions on Amino Acid Substitutions</A>
	 <LI><A HREF="#SIFT_OUTPUT_GENOME">Genome Tools</A>
	<LI><A HREF="#SIFT_OUTPUT_SINGLE">Single Protein Tools</A>
	</UL>

<LI><A HREF="#EXAMPLES">Examples</A>
	<UL>
	<LI><A HREF="#LACI">LacI</A>
	<LI><A HREF="#HIV_PROTEASE">HIV-1 Protease</A>
	<LI><A HREF="#LYSOZYME">Bacteriophage T4 lysozyme</A>
	</UL>
<LI><A HREF="#PERFORMANCE">Performance</A>
	<UL>
	<LI><A HREF="#BLOS62_PERF">Compared to BLOSUM62</A>
	<LI><A HREF="#BLOS_PERF">Compared to other BLOSUM matrices</A>
	</UL>
<LI><A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=11337480&dopt=Abstract">Reference</A>
<A NAME="SIFT"><H2>What does SIFT do?</H2></A>
SIFT is a sequence homology-based tool that <b>s</b>orts <b>i</b>ntolerant <b>f</b>rom <b>t</b>olerant amino acid 
substitutions and predicts whether an amino acid substitution in a protein will have a phenotypic effect. SIFT is based on the premise that protein evolution is correlated with protein function.  Positions important for function should be conserved in an alignment of the protein family, whereas unimportant positions should appear diverse in an alignment. 

<A NAME="SIFT_FOR_ME"><H2>What can SIFT do for me?</H2></A>
<BR>
<ol type=1>
<li>If you have a protein of interest that you would like to mutations in, submit your sequence to SIFT.  Regions that do not tolerate many substitutions will be highlighted in red in the Score output file, and you can target these regions for mutation.
</li>
<BR>
<li>If you have mutant proteins with single amino acid substitutions, SIFT will predict which mutants may have a phenotypic effect before you carry out your functional assays.
</li>
<BR>
</ol>

<A NAME="SIFT_PROCEDURE"><H2>How does SIFT work?</H2></A>
<b><u>Brief Summary</u><BR>
<BR>
SIFT takes a query sequence and uses multiple alignment information to predict tolerated and deleterious substitutions for every position of the query sequence.  SIFT is a multistep procedure that (1) searches for similar sequences, (2) chooses closely related sequences that may share similar function to the query sequence , (3) obtains the alignment of these chosen sequences, and (4) calculates normalized probabilities for all possible substitutions from the alignment.  Positions with normalized probabilities less than 0.05 are predicted to be deleterious, those greater than or equal to 0.05 are predicted to be tolerated.
</b><BR>
<BR>
<u><b><font size=+1>Procedure (the details)</font></b></u>:<BR>
<BR>
<ol type =1>
<li> <b>Get related sequences.</b>  A <A HREF="http://www.ncbi.nlm.nih.gov/BLAST/">PSI-BLAST</A> search against a database is executed on the query sequence.  
<BR>
<BR>
Parameters: 4 iterations, expectation value .0001,  e-value threshold for inclusion in multipass model 0.002<BR>
<i>Update 05/15/01</i> Number of PSI-BLAST iterations reduced to 2 to save time and prevent the search from diverging.</i><BR>
<BR>
<li>  <b>Choose closely related sequences.</b><BR>
<BR>
<b>As described in <i>Genome Research</i> 11:963-87:</b> 
 We desire to have sequences that are similar in function as well as structure to the query sequence.  To do so, we select only a subset of sequences from the PSI-BLAST results.  <BR>
<ol type=a>
<li> Group sequences found from the PSI-BLAST search that are more than 90% identical together and make a consensus sequence for each group by choosing the amino acid that occurs most frequently at each position.<BR>
<li> <A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=1689055&dopt=Abstract">MOTIF</A>
 finds conserved regions among the query sequence and the consensus sequences from (a) that were derived from at least two sequences.<BR>
<li> After the conserved regions in the query sequence have been identified by MOTIF, these regions are extracted from the sequences aligned by PSI-BLAST.<BR>
<li>The conserved regions of the query sequence and those consensus sequences more than 90% identical are converted to a PSI-BLAST checkpoint file.<BR>
<li> The checkpoint file is given to PSI-BLAST to search among the remaining 
conserved regions of the consensus sequences not included in the seed 
checkpoint file. The top hit is added to the alignment corresponding tothe seed
 checkpoint file and the <A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=3525846&dopt=Abstract">conservation</A> over the entire alignment of conserved 
regions is calculated.  If conservation does not decrease, the consensus 
sequence is added to the alignment and the checkpoint file rebuilt.  
(e) iterates until conservation decreases.
<BR>
</ol>
<center><b>OR</b></center><BR>
<A NAME="SIFT_PROCEDURE_CONSERVATION"><b>SIFT by conservation</b></A>: In the original version of SIFT, an arbitrary number of sequences is added.  In this version, sequences are continually added until they reach a sequence conservation cutoff, set by the user.<BR>
If the sequences for which prediction is based on are very diverse (low conservation cutoff),
only substitutions at the strongly 
conserved positions will be predicted as deleterious.
If the sequences chosen for prediction are very similar to each other (high conservation cutoff), then most substitutions will be predicted as deleterious.
<BR>
Users can choose the degree of sequence conservation: 
they can opt for detecting most of the deleterious substitutions 
(use a high sequence conservation) , or predict fewer deleterious substitutions but  with a high level of certainty (use a low sequence conservation).
<BR>
<ol type =a>
<li>  Group sequences found from the PSI-BLAST search (step 1) that are more than 90% identical
together and make a consensus sequence for each group by choosing the amino acid that occurs most frequently at each position.
<li>  The query sequence and its checkpoint file is given to PSI-BLAST to search
among the consensus sequences.  The top hit is added and aligned to the query sequence.  Information is calculated for each position in the alignment, and the median of these values is obtained.  If the median conservation over all positions
does not fall below a given cuttoff, the hit is retained in the alignment and the checkpoint file rebuilt.  The process repeats until the median conservations as long as the median information does not fall below the cutoff.<BR>
</li>
</ol>

<BR>

<BR>
The sequences picked from this iterative procedure are chosen as closely related sequences.  You can also <A HREF="SIFT_related_seqs_submit.html">submit</A> your own sequences.<BR>
<BR>
<li>  <b>Obtain alignment. </b>  
Since PSI-BLAST alignments are fairly
accurate and long 
<A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=10977091&dopt=Abstract">(Sauder & Dunbrack, 2000)</A>, we obtain the alignment of the sequences chosen in (2) from the initial PSI-BLAST search results (1).<BR>
You can also <A HREF="SIFT_aligned_seqs_submit.html">submit</A> your own 
alignment of your query sequence with other sequences.<BR>
<BR>
<li>  <b>Calculate probabilities. </b>
At each position of the alignment, each amino acid i appears at a frequency n<sub>i</sub>. 
Using the n<sub>i</sub>'s, the probabilities of amino acids are estimated according to Dirichlet mixtures (d<sub>i</sub>'s.
The final probability of an amino acid appearing at a position,  p<sub>i</sub>, is a weighted average of the observed frequencies and the Dirichlet estimation.  The weight of the observed frequencies is the number of sequences used to construct the alignment.
The weight of the Dirichlet estimated probabilities is an exponential function of a diversity measure (<b>Div</b>) calculated by<BR>
<BR>
<center> <b>Div</b> = SUM ( rank<sub>i</sub> * n<sub>i</sub>) </center>
<BR>where rank<sub>i</sub> is the rank amino acid i has in reference to the original amino acid when BLOSUM62 substitution scores for the original amino acid are ranked from highest to lowest. 
<BR>
Probabilities are normalized by dividing by max{Pr(amino acid)}.<BR>  
<BR>
<i>Update: 08/08/01</i>: Prior to calculating the probabilities, sequences > 90% identical to the query sequence are removed.  This eliminates the possibility that the sequence containing your substitution of interest is already represented in the database therefore and will trivially be predicted as tolerated.<BR>
</li>
<BR>
</ol>
We have found by comparison to experimental data that substitutions with less than 0.05 are deleterious.  We use this as a cutoff for prediction.  We strongly suggest users examine the normalized probabilities manually.  If your substitutions are slightly above the 0.05 cutoff, you might want to consider this as a deleterious substitution.
<BR>
</ol>
<BR>
<A NAME="SIFT_INPUT"><H2>Input for SIFT</H2></A>
You can submit a protein sequence (slow), or your query sequence along with related
 sequences (fast) or your query sequence aligned with related sequences (even faster).
<BR>
<BR>
<A NAME="SIFT_INPUT_BLINK"><H3>Submitting a NCBI GI #</H3></A>
You can <A HREF="SIFT_BLink_submit.html"> submit a NCBI GI #id </A> to obtain SIFT predictions. Predictions are based on pre-computed BLAST searches and are returned within a minute. This is the preferred method of submission. 
<BR>
<BR>
To find a NCBI GI # for a particular protein sequence, go to the <A HREF="http://www.ncbi.nlm.nih.gov/sites/entrez?db=protein">NCBI protein database</A HREF> and   

type in the gene name.  If you get back too many results you can narrow it down by specifying the organism.  For example, if looking for the human MLH1 gene, type <b>"MLH1"[GENE] AND "homo sapiens"[ORGANISM]</b> into the NCBI text box and a shorter list of genes restricted to human will be returned.
<A NAME="SIFT_INPUT_SEQUENCE"><H3>Submitting a sequence</H3></A>
You can <A HREF="SIFT_seq_submit2.html"> submit a protein sequence </A> in FASTA format.  
The entire <A HREF="#SIFT_PROCEDURE">SIFT procedure</A> will be executed and results will be returned to you.  This procedure is slow; if you have additional information about the protein, you can get your results much faster.
<BR>
<A NAME="SIFT_INPUT_SEQUENCES"><H3>Submitting a group of related sequences</H3></A>
If you know of proteins related to your query protein, you can get results much 
<b><i>faster</i></b> by <A HREF="SIFT_related_seqs_submit.html">submitting your sequence and related sequences</A>.
  Steps <b>(1)</b> & <b>(2)</b> of the <A HREF="#SIFT_PROCEDURE">SIFT procedure</A> are skipped.  Submit in FASTA format with your protein of interest as the first sequence in the file. 
<BR>
<A NAME="SIFT_INPUT_ALIGNMENT"><H3>Submitting a multiple alignment</H3></A>
<BR>
If you have a multiple alignment containing your protein of interest, you can 
<A HREF="SIFT_aligned_seqs_submit.html">submit the alignment</A> in CLUSTAL, MSF, or FASTA format.  Your protein should be first in the alignment.
The length of the alignment should correspond to the query protein and there should be no gaps in the query protein sequence.  
Since steps <b>(1)</b> through <b>(3)</b> are skipped in the <A HREF="#SIFT_PROCEDURE">SIFT procedure</A>, you will get your results <b><i>SUPER-DUPER FAST</b></i> and we encourage you to use this submission form instead of the others.  <BR>
<BR> 

<BR>
<A NAME="substitution"><H2>Submitting Substitutions</H2>
</A><BR>
SIFT will return predictions on whether your substitutions are tolerant or intolerant based on the scores. <BR> 
<BR>
The format for a substitution is to have X#Y where X is the original amino acid, # is the position of the substitution and Y is the new amino acid.  One substitution per line is allowed. <BR>
<BR>
Example: <BR>
<spacer type=horizontal size=50>M1Y<BR>
<spacer type=horizontal size=50>K3S<BR>
<spacer type=horizontal size=50>T4P<BR>
<BR>
<HR>
<A NAME="SIFT_OUTPUT"><H1>SIFT Output</H1></A>
</center></H1>

<A NAME="SIFT_OUTPUT_SUBST"><h2>SIFT Predictions for Substitutions</h2>
<A><table border="1">
<th width=100><b>Output</b></th>
<th width=300><b>Description</b></th>
</th>
<tr>
<td width=100><b>SIFT Score</b></td>
<td>Ranges from 0 to 1. The amino acid substitution is predicted damaging is the score is <= 0.05, and tolerated if the score is > 0.05. </td>
</tr>
<tr>
<td width=100><b>Median Info</b></td>
<td>Ranges from 0 to 4.32, ideally the number would be between 2.75 and 3.5.  This is used to measure the diversity of the sequences used for prediction.
A warning will occur if this is greater than 3.25 because this indicates that the prediction was based on closely related sequences.
</td>
</tr>
<tr>
<td width=100><b>Seqs at Position</b></td>
<td width=400>This is the number of sequences that have an amino acid at the position of prediction. SIFT automatically chooses the sequence for you, but if the substitution is located at the beginning or end of the protein, there may be only a few sequences represented at that position, and this column indicates this. </td>
</tr>

</table>
<HR>

<h2><A NAME="SIFT_OUTPUT_GENOME">Genome Tool Output</A></h2>
<BR>
Below is example output returned when genomic variants are submitted. <BR>
<BR>
<table border="1" cellspacing="0" cellpadding="4"><thead><tr><th>Coordinates</th><th>Codons</th><th>Transcript ID</th><th>Protein ID</th><th>Substitution</th><th>Region</th><th>dbSNP ID</th><th>SNP Type</th><th>Prediction</th><th>Score</th><th>Median Info</th><th># Seqs at position</th><th>User Comment</th></tr>

<tr>

<tr>
<td>1,100624830,1,T/A</td><td>ATA-tTA</td><td>ENST00000342895</td><td>ENSP00000344470</td><td>I121L</td><td>EXON CDS</td><td><A HREF="http://www.ncbi.nlm.nih.gov/sites/entrez?db=snp&cmd=search&term=+rs34920283" target="_blank">rs34920283:A</A></td><td>Nonsynonymous</td><td>TOLERATED</td><td>0.59</td><td>3.06</td><td>28</td></tr>
<tr>
<td>22,30163533,1,A/C</td><td>GAG-GcG</td><td>ENST00000330029</td><td>ENSP00000332887</td><td>E49A</td><td>EXON CDS</td><td><A HREF="http://www.ncbi.nlm.nih.gov/sites/entrez?db=snp&cmd=search&term=+rs11554363" target="_blank">rs11554363:C</A></td><td>Nonsynonymous</td><td><font color=red>DAMAGING</font></td><td>0.03</td><td>3.04</td><td>50</td></tr>

<tr>
<tr>
<td>X,10085674,1,T/C</td><td>GAT-GAc</td><td>ENST00000380861</td><td>ENSP00000370242</td><td>D525D</td><td>EXON CDS</td><td><A HREF="http://www.ncbi.nlm.nih.gov/sites/entrez?db=snp&cmd=search&term=+rs6530368" target="_blank">rs6530368:C</A></td><td>Synonymous</td><td>N/A</td><td>N/A</td><td>N/A</td><td>N/A</td></tr>

<tr>
<tr>
<td>21,19638426,1,T/G</td><td>TTG-gTG</td><td>ENST00000338326</td><td>ENSP00000339975</td><td>L223V</td><td>EXON CDS</td><td>novel</td><td>Nonsynonymous</td><td><font color=red>DAMAGING *Warning! Low confidence.</font></td><td>0</td><td>4.32</td><td>2</td></tr>

<tr>
<td>2,230633386,1,G/A</td><td>CAG-tAG</td><td>ENST00000283943</td><td>ENSP00000283943</td><td>Q1910*</td><td>EXON CDS</td><td><A HREF="http://www.ncbi.nlm.nih.gov/sites/entrez?db=snp&cmd=search&term=+rs1803846" target="_blank">rs1803846:A</A></td><td>Nonsynonymous</td><td>N/A</td><td>N/A</td><td>N/A</td><td>N/A</td></tr>
<tr>
<td>2,230312220,1,G/A</td><td>CCC-CtC</td><td>ENST00000341772</td><td>ENSP00000345229</td><td>P433L</td><td>EXON CDS</td><td><A HREF="http://www.ncbi.nlm.nih.gov/sites/entrez?db=snp&cmd=search&term=+rs17853365" target="_blank">rs17853365:A</A></td><td>Nonsynonymous</td><td>TOLERATED</td><td>0.11</td><td>3.02</td><td>160</td></tr>

<tr>
</tr>
</tr>

</table>
<ul>
<li>The first column indicates the variant submitted. If alleles are submitted with respect to the - strand, they will be automatically converted to + strand.  <i>Please not that if you do not submit the variant correctly, it will default to a synonymous change. One way to check is if the reference and non-reference alleles in the coordinates column now match, this indicates that you most likely did not submit your variant correctly.</i><BR>
<li>The second column denotes the codon that has been changed, the bases are with respect to + mRNA orientation.<BR> 
<li>If dbSNP has a variant overlapping at the same position, the rs ID is displayed. However, the alleles may not be the same.<BR>
<li>SIFT predictions are as described <A HREF="#SIFT_OUTPUT_SUBST">above</A>.<BR> 
</ul>
<BR>
<BR>
<A NAME="SIFT_OUTPUT_GENOME_Error"></A>
<b>Gene Annotation Error</b> :  We check that the Ensembl gene annotation  codes for the expected Ensembl protein.  For example, if an Ensembl gene is from positions 3-56 in the NCBI reference genome, we extract those DNA bases from the NCBI reference genome, translate it, and then check that it matches the corresponding Ensembl protein sequence.<BR>
If the Ensemble gene annotation does not match the expected protein sequence, we do not annotate the coding variant.  16% of the proteins from NCBI36 had this error, and 7% from NCBI37 have this error. Therefore, if you receive this error, we recommend you resubmit with NCBI37 coordinates (after <A HREF="http://genome.ucsc.edu/cgi-bin/hgLiftOver">converting NCBI36 to NCBI37</A HREF>), and if you still receive this error, we recommend that you annotate by hand. 
<BR>

<HR>
<BR>
<A NAME="SIFT_OUTPUT_SINGLE"><h2>Single Protein Output </h2></A>
<BR>For single protein submissions, the following output is also returned: <BR>
<BR>
<ol type=1>
<li><A NAME="SIFT_OUTPUT_PROB">A table</A> of probabilities   <A HREF="#SIFT_PROCEDURE">[Procedure]</A><BR>

Here is an example of one of the rows in the table.  <BR>
<BR>
<table cellspacing=0 border=0 width=0 cols=21>
<tr><th>pos</th>
<th>A</th><th>C</th><th>D</th><th>E</th><th>F</th><th>G</th><th>H</th><th>I</th>
<th>K</th><th>L</th><th>M</th><th>N</th><th>P</th><th>Q</th><th>R</th><th>S</th>
<th>T</th><th>V</th><th>W</th><th>Y</th></tr>
<tr><tr><th> 9I 0.75</th><td>0.71</td><td>0.12</td><td>0.39</td><td>0.68</td>
<td>0.35</td><td>0.36</td><td>0.30</td><td>0.81</td><td>1.00</td><td>0.87</td>
<td>0.24</td><td>0.42</td><td>0.28</td><td>0.54</td><td>0.76</td><td>0.58</td>
<td>0.58</td><td>0.94</td><td><font color=red>0.02</font></td><td>0.39</td></tr>
</table>
<BR>
This lists normalized probabilities for position <b>9I</b> of the query sequence.  
Underneath <b>9I</b> is the fraction of sequences that are represented at this 
position.  In this case, 75% of the sequences had a basic amino acid appearing 
in the sequence; 25% had either gaps or Xes.  
The normalized probability for an I->W substitution is < 0.05 
so it is predicted deleterious and highlighted in red.<BR>
<BR>
<li> <A NAME="SIFT_OUTPUT_PRED">Predictions</A> for each position<BR>
Here is an example of the output.
<table cellspacing=0
border=0 width=0 cols=41><th colspan=20 align=center>
Predict Not Tolerated</th><BR><th>Position</th>
<th>Seq Rep</th><BR><th colspan=20 align=center>
Predict Tolerated</th>
<tr><td><font color=gray> </font></td><td><font color=gray> </font></td>
<td><font color=gray> </font></td><td><font color=green>c</font></td>
<td>
<font color=black>w</font></td><td><font color=blue>d</font></td><td>
<font color=black>f</font></td><td><font color=black>m</font></td><td>
<font color=black>i</font></td><td><font color=green>y</font></td><td>
<font
color=black>v</font></td><td><font color=black>g</font></td><td><font 
color=black>p</font></td><td><font color=green>s</font></td><td><font 
color=red>h</font></td><td><font color=green>n</font></td><td><font 
color=black>
a</font></td><td><font color=black>l</font></td><td><font color=green>
t</font></td><td><font color=blue>e</font></td><th>7Q</th><td>0.95</td>
<td><font color=red>K</font></td><td><font color=green>Q</font></td><td>
<font color=red>R</font></td><td><font color=gray> </font></td><td><font
 color=gray></font></td><td><font color=gray></font></td><td><font 
color=gray></font></td><td><font color=gray></font></td><td><font 
color=gray></font></td><td><font color=gray></font></td><td><font color=gray>
</font></td><td><font color=gray></font></td><td><font color=gray>
</font></td><td><font color=gray> </font></td><td><font color=gray></font>
</td><td><font color=gray></font></td><td><font color=gray> </font></td>
<td><font color=gray> </font></td><td><font color=gray> </font></td><td>
<font color=gray> </font></td></tr>
</table>
<BR>
At position <b>7Q</b> in the query sequence, 95% of the sequences have an amino acid 
appearing at this position.  K, Q, R are predicted as tolerated and are observed in 
the alignment (capitalized).  C, W, D, F, M, I, Y, V, G, P, S, H, N, A, L, T, E are 
predicted to be deleterious because they have normalized probabilities < 0.05 and 
none of these appear in the alignment (small letters).  Amino acids are color coded: 
nonpolar,
<font color=green>uncharged polar</font>, <font color=red>basic</font>,
<font color=blue>acidic</font>. <BR>
<BR>
</li>

<li>  If you submit substitutions, in addition to returning the prediction, SIFT will return: <BR>
<spacer type=horizontal size=10>(1) The number of sequences used at the position of substitution for prediction (not counting sequences with gaps at that position) <BR>
<spacer type=horizontal size=10>(2) The median sequence information used to measure the diversity of the sequences used for prediction.<BR>
<spacer type=horizontal size=30>
<font size=-1>The median sequence information is calculated by first calculating the 
<A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed
&list_uids=3525846&dopt=Abstract">information</A>
<BR>
<spacer type=horizontal size=30>at each position 
and then obtaining the median over all
positions.</font>
<BR> 
<BR>

<b>  We recommend that
a substitution predicted to be deleterious with median sequence information
greater than 3.25 and especially those greater than 3.5 be
taken with caution </b>
because this prediction was based on closely related sequences. The
substitution may be at a position that has not had time to evolve and is
conserved by
chance (and hence predicted to be deleterious) when 
the position can actually tolerate substitutions and given more evolutionary
time, will eventually mutate to different amino acids. 
<BR>
<i>Exception:</i> If your protein belongs to a protein family that is especially 
conserved (such as histones) then using a higher median sequence information 
may be fine.<BR>
<BR>

	Sample output:<BR>
<BR>
Substitution at pos 2 from S to F is predicted to be <font color=red>DELETERIOUS</font> with a score of 0.01.<BR>
<spacer type=horizontal size=5>Median sequence information: 3.44 <BR>
<spacer type=horizontal size=5>Sequences represented at this position:3 <BR>
<font color=red>WARNING!!</font>  This substitution may have been predicted as deleterious just because <BR> 
 the prediction was based on sequences too closely related.  We recommend a median<BR> 
 sequence information <= 3.25 for reasonable accuracy and for which sequence 
 diversity is adequate.<BR>
<BR> 
<BR>
Substitution at pos 60 from E to L is predicted to be <font color=red>DELETERIOUS</font> with a score of 0.00.<BR>
<spacer type=horizontal size=5>    Median sequence information: 2.72 <BR>
<spacer type=horizontal size=5>    Sequences represented at this position:116 <BR>
<BR>
Explanation:<BR>
<spacer type=horizontal size=10>The substitution S2F was predicted to be deleterious.  However, a warning was given because the sequences used for prediction were not very diverse from each other. 
<BR>
<BR>
E60L was predicted to be deleterious and had a low median sequence information, indicating that the sequences used for prediction were diverse enough . 

<BR>

<BR>
<BR>

 
</ol>

<HR>
<A NAME="CHANGING_PARAMETERS"><H2>Changing Parameters</H2></A>
<A NAME="TIMING OUT">
<b>If SIFT is timing out:</b><BR>
<BR>
<UL>
<LI><i>Check the length of your protein.</i><BR>
The duration SIFT runs is proportional to the length of the protein.
If your protein is very long (more than 500 amino acids), truncate your protein
to the region you're intersted in (a particular domain or region carrying the substitution) and rerun SIFT.  The optimal length is 300-500 amino acids.  Don't forget to shift the location of your substitution.<BR>
<BR>
<LI><i>The server is busy.</i>  In this case, change the default parameters so that your 
query takes less time to run and will be completed before the time limit.
You can:<BR> 
<BR>
<ol type=1>
<li> change median conservation of sequences to 3.00 (from default 2.75).<BR>
-OR-<BR>
<li> change the database to SWISS-PROT.<BR>
</ol><BR>
If changing these parameters will affect the quality of prediction, a 
warning will be given with the predictions of the substitutions.
 <BR>
<A NAME="CHOOSING_DATABASE">
</ul>
<BR>
<b>If you are getting the error "not enough sequence information for prediction" (i.e. SIFT returns error or median sequence information > 3.25) :</b>  <BR>
<ul><li>Increase the database size.</font><BR>
<BR>
<u>Database options</u><BR>
<BR>
<spacer type=horizontal size=15><b>SWISS-PROT</b> 
= small but high quality. 
Less likely to time out but may not have enough sequences for prediction.<BR>
<spacer type=horizontal size=15><b>SWISS-PROT/TrEMBL</b> 
= larger than SWISS-PROT, good quality <b><- what I like to use</b><BR>
<spacer type=horizontal size=15><b>nr (nonredundant database) </b>
= largest database, but not that good quality. Increases possibility of timing out but can also give more sequence information. <BR>
<BR>
If there is not enough sequences to make your prediction  (getting a warning that the median sequence information > 3.25 for your substitution), increase the size of the database.  
Change the database option from SWISS-PROT/TrEMBL (default)
to the nonredundant database.<BR>
</ul><BR>
<HR>

<A NAME="EXAMPLES"><H1>Examples</H1></A>

<A NAME="LACI"><H2>LacI</H2></A>
<H3>Submission</H3>
<i>Paste in Sequence</i><BR>
<PRE>
>gi|2506562|sp|P03023|LACI_ECOLI LACTOSE OPERON REPRESSOR
MKPVTLYDVAEYAGVSYQTVSRVVNQASHVSAKTREKVEAAMAELNYIPNRVAQQLAGKQSLLIGVATSS
LALHAPSQIVAAIKSRADQLGASVVVSMVERSGVEACKAAVHNLLAQRVSGLIINYPLDDQDAIAVEAAC
TNVPALFLDVSDQTPINSIIFSHEDGTRLGVEHLVALGHQQIALLAGPLSSVSARLRLAGWHKYLTRNQI
QPIAEREGDWSAMSGFQQTMQMLNEGIVPTAMLVANDQMALGAMRAITESGLRVGADISVVGYDDTEDSS
CYIPPLTTIKQDFRLLGQTSVDRLLQLSQGQAVKGNQLLPVSLVKRKTTLAPNTQTASPRALADSLMQLA
RQVSRLESGQ
</PRE>
<i>With Parameters:</i>
<dl>
<dt>Database:                 <b>SWISS-PROT/TREMBL</b><BR>
<dt>Account for gaps?:    <b>YES</b><BR>
</dl>
<H3>SIFT Results</H3>
A PSI-BLAST <A HREF="LacI_alignment_SPTR.html">alignment</A> with 27 sequences
were returned along
with 
<A HREF="LacI_SPTRprob.html">normalized probabilities</A>.
<BR>
When results from tblastn on the microbial genomes were added, 54 sequences were chosen.  This <A HREF="LacI_alignment_SPTRgenomes.html">alignment</A> gave similar predictions and <A HREF="LacI_SPTRgenomesprob.html">probabilities</A>.  These probabilities were used for to test <A HREF="#PERFORMANCE">SIFT performance</A>.
<BR>

<A NAME="HIV_PROTEASE"><H2>HIV-1 Protease</H2></A>
<H3>Submission</H3>
<i>Paste in Sequence</i><BR>
<PRE>
>HIV-1 protease
PQITLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMSLPGRWKPKMIGGI
GGFIKVRQYDQILIEICGHKAIGTVLVGPTPVNIIGRNLLTQIGCTLNF
</PRE>
<i>With Parameters:</i>
<dl>
<dt>Database:                 <b>SWISS-PROT</b><BR>
<dd><i>SWISS-PROT was chosen rather than SWISS-PROT/TREMBL or the nonredundant database because there may be nonfunctional proteins in the latter databases that would corrupt our predictions.</i><BR>
<BR>
<dt>Account for gaps?:	  <b>YES</b><BR>
</dl>
<H3>SIFT Results</H3>
A PSI-BLAST <A HREF="HIV_alignment.html">alignment</A> with 39 sequences 
was returned along 
with <A HREF="HIV_predictiontable.html">predictions </A>
and <A HREF="HIV_scores.html">probabilities</A>.
<BR>

<A NAME="LYSOZYME"><H2>Bacteriophage T4 lysozyme</H2></A>
<H3>Submission</H3>
<i>Paste in Sequence</i><BR>
<PRE>
>LYCV_BPT4|P00720
MNIFEMLRIDEGLRLKIYKDTEGYYTIGIGHLLTKSPSLNAAKSELDKAIGRNCNGVITKDEAEKLFNQDVDAAVRGI
LRNAKLKPVYDSLDAVRRCALINMVFQMGETGVAGFTNSLRMLQQKRWDEAAVNLAKSRWYNQTPNRAKRVITTFRTGTW
DAYKNL
</PRE>
<i>With Parameters:</i>
<dl>
<dt>Database:                 <b>SWISS-PROT/TR</b><BR>
<dt>Account for gaps?:    <b>YES</b><BR>
</dl>
<H3>SIFT Results</H3>
An <font color=red>ERROR</font> was returned with the following message:<BR>
<BR>
The query sequence did not share any conserved regions with the sequences 
found in PSI-BLAST.  This may be due to divergence of PSI-BLAST results.  We 
suggest you run a PSI-BLAST search on your own and hand-pick sequences that
may be related to your sequence.<BR>
<BR>
From the <A HREF="http://www.ncbi.nlm.nih.gov/BLAST/">PSI-BLAST</A> search 
, the <A HREF="Lysozyme_gappedaln.html">alignment</A> that results is severely gapped.  
Only one other sequence, VG05_BPT4, aligns well with the query sequence.
Submitting the pairwise alignment to <A HREF="SIFT_aligned_seqs_submit.html">SIFT</A>, <A HREF="lysozymepred.html">predictions </A>
and <A HREF="lysozymeprob.html">normalized probabilities</A> are returned.<BR>
Remarkably, with only two sequences, SIFT <A HREF="#PERFORMANCE">performs well</A> in predicting deleterious substitutions.
<BR>

<A NAME="PERFORMANCE"><H1>Performance</H1></A>
To test SIFT against experimental data, we chose unbiased data sets in which mutagenesis was performed throughout the entire protein and both positive and 
negative phenotypes were assayed.  
 Phenotypes with weakened activity were grouped with loss of function 
phenotype and SIFT was tested for its ability to predict these substitutions as deleterious.
As a control, the BLOSUM62 substitution scoring matrix is used where nonnegative scores are predicted as tolerated; negative scores as deleterious.<BR>
<BR>
<i>Measures we used:</i>
<dl>
	<dt><b>Tolerant prediction accuracy</b></dt>
		<dd>(# of tolerant substitutions correctly predicted) / tolerated</dd>
	<dt><b>Deleterious prediction accuracy</b></dt>
		<dd>(# of deleterious substitution correctly predicted) / deleterious</dd>
	<dt><b>Total Prediction Accuracy/b></dt>
		<dd>(total # of correctly predicted substitutions) / (total # of substitutions)<BR>
	<dt><b>Experimental Prediction Accuracy</b></dt>
		<dd>(redicted deleterious <i>and</i> phenotypically deleterious) / (# predicted deleterious)</dd>

</dl>

<A NAME="BLOS62_PERF"><center>SIFT performance compared to BLOSUM62</center></A>

<table border=1 cellspacing=5>
<tr align=center>
<th>Test Case</th><th>Method</th><th>Tolerant<BR> Prediction<BR> Accuracy</th>
<th>Deleterious<BR> Prediction<BR> Accuracy</th>
<th>Total<BR>Prediction<BR>Accuracy</th>
<th>Experimental<BR> Prediction<BR> Accuracy</th>
</tr>

<tr align=center>
<th rowspan=2><A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=8794873&dopt=Abstract">LacI</A></th>
<th>SIFT</th><td>78%<BR>(1747/2254)</td>
<td>57%<BR>(989/1750)<BR></td><td><b>68%</b><BR>(2736/4004)</td>
<td><b>66%</b><BR>(989/1496)</td>
</tr>
<tr align=center>
<th>BLOSUM62</th><td>31%<BR>(696/2254)</td><td>84%<BR>(1475/1750)
</td><td>54%<BR>(2171/4004)</td><td>49%<BR>(1475/3033)</td>
</tr>


<tr align=center>
<th rowspan=2><A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=2666861&dopt=Abstract">HIV-1 protease</A></th>
<th>SIFT</th><td>70%<BR>(78/111)</td><td>82%<BR>(184/225)<BR></td>
<td><b>78%</b><BR>(262/336)</td><td><b>85%</b><BR>(184/215)</td>
</tr>
<tr align=center>
<th>BLOSUM62</th><td>63%<BR>(70/111)</td><td>73%<BR>(165/225)</td>
<td>70%<BR>(235/336)</td><td>80%<BR>(165/206)</td></tr>

<tr align=center>
<th rowspan=2><A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=1942069&dopt=Abstract">Bacteriophage T4 lysozyme</A></th>
<th>SIFT</th><td>59%<BR>(817/1377)</td><td>72%<BR>(460/638)<BR></td>
<td><b>63%</b><BR>(1277/2015)</b>
</td><td><b>45%</b><BR>(460/1020)</td>
</tr>

<tr align=center>
<th>BLOSUM62</th><td>30%<BR>(406/1377)</td><td>85%<BR>(542/638)</td>
<td>47%<BR>(948/2015)</td><td>36%<BR>(542/1513)
</td></tr>

</table>
<BR>

In all test cases, SIFT predicts better than BLOSUM62.  SIFT always has a higher
 number of correctly predicted substitutions than BLOSUM62 (positive gain). Also,
 for all test sets, the number of substitutions predicted by SIFT to alter
protein function is reduced.  For a biologist investigating substitutions
predicted to have a deleterious effect, fewer functional assays need to be
performed and a higher proportion of the assays will yield affected phenotypes.
<BR>
<BR>

<A NAME="BLOS_PERF">Performances of other matrices.</A>
 Results were similar to BLOSUM62.
<table border=1 cellspacing=5>
<tr align=center>
<th>Test Case</th><th>Method</th><th>Tolerant<BR> Prediction<BR> Accuracy</th>
<th>Deleterious<BR> Prediction<BR> Accuracy</th>
<th>Total<BR>Prediction<BR>Accuracy</th>
<th>Experimental<BR> Prediction<BR> Accuracy</th>
</tr>

<tr align=center>
<th rowspan=2><A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=8794873&dopt=Abstract">LacI</A></th>
<th>BLOSUM45</th><td>32%<BR>(729/2254)</td><td>81%<BR>(1421/1750)</td>
<td>54%<BR>(2150/4004)<td>48%<BR>(1421/2946)</td></tr>
<tr align=center>
<th>BLOSUM80</th><td>23%<BR>(526/2254)</td><td>88%<BR>(1547/1750)</td>
<td>52%<BR>(2073/4004)<td>47%<BR>(1547/3275)</td></tr>


<tr align=center>
<th rowspan=2><A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=2666861&dopt=Abstract">HIV-1 protease</A></th>
<th>BLOSUM45</th><td>64%<BR>(71/111)</td><td>72%<BR>(161/225)</td>
<td>69%<BR>(232/336)</td><td>80%<BR>(161/201)</td></tr>
<tr align=center>
<th>BLOSUM80</th><td>59%<BR>(65/111)</td><td>76%<BR>(170/225)</td>
<td>70%<BR>(235/336)</td><td>78%<BR>(170/216)</td></tr>

<tr align=center>
<th rowspan=2><A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=1942069&dopt=Abstract">Bacteriophage T4 lysozyme</A></th>
<th>BLOSUM45</th><td>31%<BR>(430/1377)</td><td>82%<BR>(522/638)</td>
<td>47%<BR>(952/2015)<td>36%<BR>(542/1469)</td></tr>

<tr align=center>
<th>BLOSUM80</th><td>22%<BR>(301/1377)</td><td>91%<BR>(577/638)</td>
<td>44%<BR>(878/2015)</td><td>35%<BR>(577/1653)</tr>

</table>


<HR>
Page last modified <MODIFICATION_DATE>August 2001</MODIFICATION_DATE>
<Address>
Questions or comments?<BR>
Contact <A HREF="/sift-bin/contact.pl">us</A>
</Address>


