<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<title>SIFT Homepage</title>
<link rel="stylesheet" type="text/css" href="/www/css/main.css" />
</head>
<!-- Begin Wrapper -->
   <div id="wrapper">
    <!-- Begin Header -->
         <div id="header">



<body bgcolor=white>
<TITLE>SIFT: Related Sequences Submission</TITLE>
<H1><img src="pics/sifter.gif" align=left width="15%">
S<font color=red>I</font>F<font color=blue>T</font>: Submit Query Sequence with Related Sequences</H1>
<BR>
<FORM ENCTYPE="MULTIPART/FORM-DATA" METHOD="POST" ACTION="/sift-bin/SIFT_related_seqs_submit.pl">
Submit your protein query sequence along with related sequences (not necessarily aligned) to SIFT.  Your query sequence is the sequence that you would like prediction on (i.e. the sequence in which you have introduced the amino acid substitutions).  
The query sequence must be first among the related sequences you enter.</b>  The protein sequences will be aligned by PSI-BLAST.
If your protein sequences are already aligned, <A HREF="SIFT_aligned_seqs_submit.html">submit your aligned sequences</A>.

<A HREF="/">[SIFT home page]</A>
<A HREF="SIFT_help.html">[More info.]</A><BR>
<BR>
</div>

<div id="content">
<font color=blue>Enter protein sequences</font>
<BR>
Enter your protein query sequence along with related sequences (not necessarily aligned).  These sequences are assumed to be functional, and is what SIFT will base its prediction on.  The normal (functional) query sequence should appear first.
<BR>
<BR>
Enter the name of the file containing the functional protein sequences <A HREF="http://www.ncbi.nlm.nih.gov/BLAST/fasta.shtml">[fasta format]</A>.
<INPUT SIZE=50 NAME="seqs_file" TYPE="file"><BR><BR>

<center>-or-</center><BR>
Paste in the functional protein sequences <A HREF="http://www.ncbi.nlm.nih.gov/BLAST/fasta.shtml">[fasta format]</A>.<BR>
<TEXTAREA NAME = "seqs" ROWS=10 COLS=70></TEXTAREA>
<BR>
<BR>
<HR>
<BR>
<font color=blue>Substitutions to be predicted on</font><BR>
Enter the substitutions to be predicted on that occur in the query sequence.<BR>
<BR>
Enter the name of file containing substitutions of interest
<A HREF="SIFT_help.html#substitution">[format]</A>:
<INPUT SIZE=50 NAME="substitution_file" TYPE="file"><BR><BR>
<center>-or-</center><BR>
Enter the substitutions of interest
<A HREF="SIFT_help.html#substitution">[format]</A>: 
<BR>
<TEXTAREA NAME = "substitutions" ROWS=5 COLS=50></TEXTAREA>
<BR><BR>
<HR>
<font color=blue>Parameters</font><BR>
<BR>
<!--<input type=checkbox checked name=gap_option value="TRUE">Account for Gaps
<BR>
<i>  If the alignment turns out to be especially gapped, resubmit with this option turned off.</i><BR>
<BR>
-->
Remove sequences more than
<input type=text name=seq_identity_filter size=3 maxlength = 3 value="100">
percent identical to query<BR>
<i>This cleans out any polymorphic alleles (sequences that
are nearly identical to your
protein but already containing the substitutions of interest).
The contaminating sequences may cause the substitution to be predicted
as tolerated.
</i>
<BR>
<HR>
<BR>
<INPUT TYPE=submit>
<INPUT TYPE=reset>
</div>
<div id="footer">
[<A HREF="/">SIFT home page</A>]
<A HREF="SIFT_help.html">SIFT help page</A>]
<BR>
<i>Questions or comments?<BR>
Contact <A HREF="/sift-bin/contact.pl"> us </A><BR>
Last updated: November 2008</i><BR>
</div>
</div>
