<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
  <head>
  <script type="text/javascript">

    // chromosome lists
    var chrs = new Array();


    chrs['Select Organism'] = ['Select Chromosome'];
    function setchrs(){

       cntrySel = document.getElementById('organism');
       chrsel = document.getElementById('chromosome');
       chrList = chrs[cntrySel.value];
       changeSelect(chrsel, chrList);
    }

    //**************************************************************************//
    // FUNCTION changeSelect(fieldObj, valuesAry, [optTextAry], [selectedVal])
    //
    //**************************************************************************//
    function changeSelect(fieldObj, valuesAry, optTextAry, selectedValue) {

      //Clear the select list
      fieldObj.options.length = 0;

      //Set the option text to the values if not passed
      optTextAry = (optTextAry)?optTextAry:valuesAry;

      //Itterate through the list and create the options
      for (i in valuesAry) {
        selectFlag = (selectedValue && selectedValue==valuesAry[i])?true:false;
        fieldObj.options[fieldObj.length] = new Option(optTextAry[i], valuesAry[i], false, selectFlag);
      }

    }
</script>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-16" />
    <!--[if IE]>
      <style type="text/css">
        #content { overflow: hidden; width: 933px; } 
        #main_content_3c #content, #main_content_2c #content { width: 953px; } 
        #main_content_2c #content #middle_content { width: 710px; }
        #main_content_3c #middle_content { width: 515px; }
      </style>
    <![endif]-->
    <!--[if lt IE 7.]>
      <style type="text/css">
        #main_logo { height: 79px; }
      </style>
    <![endif]-->
    <title>SIFT Genome</title>

    <style type="text/css">
      #middle_content_template { 
        margin: 0;
        margin-left: 210px;
      } /* end #middle_content_template */
      
      #middle_content_template, #left_content ul { text-align: left; }
    </style>

    
    
  </head>
  <body> 
    <div id="logo_search" class="clearfix">
        <div id="project">SIFT Genome
	</div>

      </div>
    </div>
    <div id="content">
      <div id="container_1">
        <div id="container_2">

          <div id="content_header">
            <div class="clearfix">

	      <FORM ENCTYPE="MULTIPART/FORM-DATA" METHOD="POST" ACTION="/sift-bin/SIFT_feed_to_chr_coords.pl">
              <div id="middle_content_template">
                  <p>This page provides SIFT predictions for a list of <b>chromosome positions and alleles</b>.
<BR><BR>Please provide your email address and input your data by using the 'upload file' option if input size is greater than 1000 chromosome locations.
			<BR><BR>
			<i>Results are deleted after an hour, so please save them! </i>

		  </p>
		  <HR>
		  <p class="header2"><b>User Input</b></p>
		  <p>
			<select name="organism" id="organism" onchange="setchrs();">
			<option value="Human_db">Homo sapiens NCBI 36</option>
			<option value="Human_db_37">Homo sapiens NCBI 37</option>
  			 <option value="Select Organism">Select Organism</option>
			</select>

			<BR>

			<BR>

			Chromosome Coordinates<BR>
			<b>Paste in comma separated list of chromosome coordinates, orientation (1,-1) and alleles</b> see [<A HREF="chr_coords_example.html">sample format</A>]<BR>
			<TEXTAREA NAME="CHR" ROWS=15 COLS=50></TEXTAREA>
			<BR>
			<BR>-or-<BR>
			<BR>
			Upload file containing chromosome coordinates and nucleotide substitutions (size limit: 1M)<BR>
			<INPUT SIZE=50 NAME="CHR_file" TYPE="file">
			</TD>
			</TR>
			</TBODY></TABLE>
			<BR>
			<HR>
			<BR>
			Enter your email address if you want the results through email : <BR>
                        <i>Please check that your address is correct and your mailbox is not full.</i><BR>

                        <INPUT SIZE=50 NAME="address">
                        </INPUT SIZE>
                        <BR>

                        <HR>

			<p class="header2"><b>Output Options</b></p><BR>
			Include the following fields in the output table<BR>
			<input type="checkbox" name="oo1" value="1" > Ensembl Gene ID<br>
			<input type="checkbox" name="oo2" value="1" > Gene Name<br>
			<input type="checkbox" name="oo3" value="1" > Gene Description<br>
			<input type="checkbox" name="oo4" value="1" > Ensembl Transcript Status (Known / Novel)<br>
			<input type="checkbox" name="oo5" value="1" > Ensembl Protein Family ID<br>
			<input type="checkbox" name="oo6" value="1" > Ensembl Protein Family Description<br>
			<input type="checkbox" name="oo7" value="1" > Protein Family Size<br>
			<input type="checkbox" name="oo8" value="1" > Ka/Ks (Human-mouse)<br> 
			<input type="checkbox" name="oo9" value="1" > Ka/Ks (Human-macaque)<br>
			<input type="checkbox" name="oo10" value="1" > OMIM Disease<br>
			<input type="checkbox" name="oo11" value="1" > Allele Frequencies (All Hapmap Populations - weighted average)<br>
			<input type="checkbox" name="oo12" value="1" > Allele Frequencies (CEU Hapmap population)<br>
			<HR><BR>
			<INPUT TYPE=submit>
			<INPUT TYPE=reset>

		  </p>
		  </FORM>
              </div>
            </div>
          </div>

        </div>
      </div>
    </div>
    <div id="home_border">
      <div id="home_bar">
        <div id="home_container">
        </div>
      </div>
    </div>

  </body>
</html>

