<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-16" />
    <!--[if IE]>
      <style type="text/css">
        #content { overflow: hidden; width: 933px; } 
        #main_content_3c #content, #main_content_2c #content { width: 953px; } 
        #main_content_2c #content #middle_content { width: 710px; }
        #main_content_3c #middle_content { width: 515px; }
      </style>
    <![endif]-->
    <!--[if lt IE 7.]>
      <style type="text/css">
        #main_logo { height: 79px; }
      </style>
    <![endif]-->
    <title>SIFT Alignment</title>

    <style type="text/css">
      #middle_content_template { 
        margin: 0;
        margin-left: 210px;
      } /* end #middle_content_template */
      
      #middle_content_template, #left_content ul { text-align: left; }
    </style>

    
    
  </head>
  <body> 
        <div id="project">SIFT Aligned Sequences 
	</div>

      </div>
    </div>
    <div id="content">
      <div id="container_1">
        <div id="container_2">

          <div id="content_header">
            <div class="clearfix">

	      <FORM ENCTYPE="MULTIPART/FORM-DATA" METHOD="POST" ACTION="/sift-bin/SIFT_aligned_seqs_submit.pl">
              <div id="middle_content_template">
                  <p><b>This tool provides SIFT predictions for a given protein FASTA sequence</b> by using aligned sequences provided by the user. Your query sequence is the sequence that you would like prediction on (i.e. the sequence in which you have introduced the amino acid substitutions).
<br>
The query sequence must be first among the aligned sequences you enter.</b> 
Partial sequences should be flanked by Xes at the beginning and end of the sequence so that those positions are not considered gaps (and hence have liberal prediction).

		  </p>
		  <HR>
		  <p class="header2"><b>User Input</b></p>
		  <p>
  		Enter your protein query sequence along with related sequences (not necessarily aligned).  These sequences are assumed to be functional, and is what SIFT will base its prediction on.  The normal (functional) query sequence should appear first.
			<BR>

			<BR>
			Name of file containing protein query sequences
			<A HREF="http://www.ncbi.nlm.nih.gov/BLAST/fasta.shtml">(fasta format)</A>.<BR>
			<INPUT SIZE=50 NAME="alignment_file" TYPE="file"><BR><BR>
			<center>-or-</center><BR>
			Paste in your protein query sequences <A HREF="http://www.ncbi.nlm.nih.gov/BLAST/fasta.shtml">(fasta format)</A>.<BR>
			<TEXTAREA NAME = "aligned_seqs" ROWS=10 COLS=70></TEXTAREA>
			<BR>
			<BR>
			<HR>
			<BR>
			Enter the substitutions of interest <A HREF="SIFT_help.html#substitution">[format]</A>:<BR>
			<TEXTAREA NAME = "substitutions" ROWS=5 COLS=50></TEXTAREA>
			<br>-or-<BR>

			<BR>
			Upload a file containing substitutions of interest <A HREF="SIFT_help.html#substitution">[format]</A>:<br>
                        <INPUT SIZE=50 NAME="substitution_file" TYPE="file"><BR><BR>

			<HR>
			<BR>
			<font color=blue>Parameters</font><BR>
			<BR>
			Remove sequences more than
			<input type=text name=seq_identity_filter size=3 maxlength = 3 value="100">
			percent identical to query<BR>
			<i>This cleans out any polymorphic alleles (sequences in the database that are nearly identical to your protein but already containing the substitutions of interest). The contaminating sequences may cause the substitution to be predicted as tolerated.</i>
			<HR>
			<BR>
			<INPUT TYPE=submit>
			<INPUT TYPE=reset>


		  </p>
		  </FORM>
              </div>
            </div>
          </div>

        </div>
      </div>
    </div>
    <div id="home_border">
      <div id="home_bar">
        <div id="home_container">
        </div>
      </div>
    </div>

  </body>
</html>

