<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>

<script type="text/javascript">

// chromosome lists
var chrs = new Array();


chrs['Select Organism'] = ['Select Chromosome'];
function setchrs(){

  cntrySel = document.getElementById('organism');
  chrsel = document.getElementById('chromosome');
  chrList = chrs[cntrySel.value];
  changeSelect(chrsel, chrList);
}

//**************************************************************************//
// FUNCTION changeSelect(fieldObj, valuesAry, [optTextAry], [selectedVal])
//
//**************************************************************************//
function changeSelect(fieldObj, valuesAry, optTextAry, selectedValue) {

  //Clear the select list
  fieldObj.options.length = 0;

  //Set the option text to the values if not passed
  optTextAry = (optTextAry)?optTextAry:valuesAry;

  //Itterate through the list and create the options
  for (i in valuesAry) {
    selectFlag = (selectedValue && selectedValue==valuesAry[i])?true:false;
    fieldObj.options[fieldObj.length] = new Option(optTextAry[i], valuesAry[i], false, selectFlag);
  }

}

</script>




<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<title>SIFT Homepage</title>
<link rel="stylesheet" type="text/css" href="/www/css/main.css" />
</head>
<!-- Begin Wrapper -->
   <div id="wrapper">
    <!-- Begin Header -->
         <div id="header">




<body onload="setchrs();" bgcolor=white>
<TITLE>SIFT Genome Analysis</TITLE>
<H1>
<img src="pics/sifter.gif" align=left width="15%">
S<font color=red>I</font>F<font color=blue>T</font> Genome 
</H1>
<FORM ENCTYPE="MULTIPART/FORM-DATA" METHOD="POST" ACTION="/sift-bin/SIFT_feed_to_chr_coords.pl">
<BR><b>SIFT Genome</b> provides SIFT predictions for a list of chromosome positions and alleles classified as SNPs.</A> This tool is based on NCBI 36 assembly of the human genome. 
<BR>
<BR>
Please provide your email address and input your data by using the 'upload file' option if input size is greater than 1000 chromosome locations. 
To avoid overloading our resources, please restrict your input file size to be smaller than 1M.
In order to do this we advice that the input may be filtered to retain chromosome locations only for coding exons
, as SIFT does not predict for non-coding locations.<BR>


<BR>
<i>Results are deleted after an hour, so please save them! </i>
<BR>
[<A HREF="/">SIFT home page</A>]
[<A HREF="SIFT_help.html">SIFT help</A>]
</A>
</div>
<div id="content">
Enter your email address if you want the results through email : <BR>
<i>Please check that your address is correct and your mailbox is not full.</i><BR>
 <BR>
<INPUT SIZE=50 NAME="address">
</INPUT SIZE>
<BR>
<BR>
<HR>
<select name="organism" id="organism" onchange="setchrs();">
  <option value="Homo_sapien">Homo_sapien</option>
  <option value="Select Organism">Select Organism</option>
</select>

<BR>

<BR>
-and-
<BR>
<BR>

<font color=blue>Chromosome Coordinates</font><BR>
One row id, set of chromosome coordinates (begin,end), orientation, nucleotide substitution per line.<BR>
[<A HREF="chr_coords_example.html">Sample format</A>]<BR></font>
<BR>


Paste in tab separated list of chromosome coordinates, orientation (1,-1) and nucleotide substitution <BR>
<TEXTAREA NAME="CHR" ROWS=5 COLS=30></TEXTAREA>
<BR>
<BR>-or-<BR>
<BR>
Upload file containing chromosome coordinates and nucleotide substitutions (size limit: 1M)<BR>
<INPUT SIZE=50 NAME="CHR_file" TYPE="file">
</TD>
</TR>
</TBODY></TABLE>
<BR>
<HR>
<BR>
<INPUT TYPE=submit>
<INPUT TYPE=reset>
</div>
<div id="footer">
[<A HREF="/">SIFT home page</A>]
<A HREF="SIFT_help.html">SIFT help page</A>]
<BR>
<i>Questions or comments?<BR>
Contact <A HREF="/sift-bin/contact.pl"> us </A><BR>
Last updated: November 2008</i><BR>
</div>
</div>
