<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-16" />
    <!--[if IE]>
      <style type="text/css">
        #content { overflow: hidden; width: 933px; } 
        #main_content_3c #content, #main_content_2c #content { width: 953px; } 
        #main_content_2c #content #middle_content { width: 710px; }
        #main_content_3c #middle_content { width: 515px; }
      </style>
    <![endif]-->
    <!--[if lt IE 7.]>
      <style type="text/css">
        #main_logo { height: 79px; }
      </style>
    <![endif]-->
    <title>SIFT BLink</title>

    <style type="text/css">
      #middle_content_template { 
        margin: 0;
        margin-left: 210px;
      } /* end #middle_content_template */
      
      #middle_content_template, #left_content ul { text-align: left; }
    </style>

    
    
  </head>
  <body> 
	SIFT BLink
    <div id="content">
      <div id="container_1">
        <div id="container_2">

          <div id="content_header">
            <div class="clearfix">

	      <FORM ENCTYPE="MULTIPART/FORM-DATA" METHOD="POST" ACTION="/sift-bin/SIFT_BLink_submit.pl">
              <div id="middle_content_template">
                  <p><b>SIFT BLink</b> provides SIFT predictions for a single protein for a given RefSeq ID or a gi number. Either all BLAST hits can be used in the protein alignment or the top hit to each organism. 
			<BR><BR>
			This procedure is much faster than having us <A HREF="SIFT_seq_submit2.html">do the search</A>.  You can also <A HREF="SIFT_related_seqs_submit.html">submit your protein sequence and related sequences</A> or <A HREF="SIFT_aligned_seqs_submit.html">aligned sequences </A> if you already have them.<BR>

		  </p>
		  <HR>
		  <p class="header2"><b>User Input</b></p>
		  <p>
			Enter your email address if you want the results through email : <BR>
			<i>Please check that your address is correct and your mailbox is not full.</i><BR>
 			
			<INPUT SIZE=50 NAME="address">
			</INPUT SIZE>
			<BR>
			
			<HR>
			<font color=blue>Protein sequence ID</font><BR>
			<BR>
			<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=protein">NCBI</A HREF> GI number OR RefSeq ID:<INPUT SIZE=20 NAME ="gi_number"></INPUT SIZE><BR></A>
			(Example: Type in 22209009 or NP_665861 to get the <A HREF="http://www.ncbi.nlm.nih.gov/sutils/blink.cgi?pid=22209009">BLink sequences<A HREF> for <A HREF="http://www.ncbi.nlm.nih.gov/entrez/viewer.fcgi?val=22209009&db=protein">gi:22209009</A>)<BR> 
			<A HREF="http://blocks.fhcrc.org/sift/SIFT_help.html#SIFT_INPUT_BLINK">How to find the GI number for your protein</A>
			<BR>
			<BR>
			<HR>
			<BR>
			<font color=blue>Substitutions to be predicted on (can be left blank)</font><BR>
			The positions of the substitutions must be relative to the protein sequence of the gi entered above.<BR>
			<BR>
			<TABLE WIDTH=75% cellpadding="15">
			<TBODY>
			<TR>
			<TD>Enter the substutions of interest <A HREF="SIFT_help.html#substitution">[format]</A>:<BR>
			<TEXTAREA NAME="substitutions" ROWS=5 COLS=30></TEXTAREA>
			</TD>

			<TD ALIGN="center" VALIGN="center" WIDTH=1000>or</TD>
			<TD ALIGN="left" VALIGN="center">enter the filename containing substitutions of interest 
			<INPUT SIZE=50 NAME="substitution_file" TYPE="file">
			</TD>	
			</TR>
			</TBODY></TABLE>
			<BR><BR>
			<HR>
			<BR>
			<font color=blue>Parameters</font><BR>
			<BR>
			Select sequences to include:
  			<SELECT NAME="sequences_to_select" SIZE="1">
  			<OPTION VALUE="ALL">all BLAST hits
  			<OPTION VALUE="BEST" SELECTED>best BLAST hit to each organism
  			</SELECT>
			<BR>
			<i>If you receive a warning that the sequences were not diverse enough for prediction on your substitution, try "all BLAST hits".  If that fails, you can submit <A HREF="SIFT_seq_submit2.html">here</A> and have SIFT find the appropriate number of sequences.</i><BR>
			<BR>
			Remove sequences more than
			<input type=text name=seq_identity_filter size=3 maxlength = 3 value="90">
			percent identical to query<BR>
			<i>This cleans out any polymorphic alleles (sequences in the database that are nearly identical to your protein but already containing the substitutions of interest). The contaminating sequences may cause the substitution to be predicted as tolerated.</i>
			<BR>
			<HR>
			<BR>
			<INPUT TYPE=submit>
			<INPUT TYPE=reset>


		  </p>
		  </FORM>
              </div>
            </div>
          </div>

        </div>
      </div>
    </div>
    <div id="home_border">
      <div id="home_bar">
        <div id="home_container">
        </div>
      </div>
    </div>

  </body>
</html>

